File: bindings/python/python/maplibre_native/resource.py
"""Resource request, response, transform, and provider APIs."""

from __future__ import annotations

from ._enum import NativeIntEnum, UnknownIntEnum
from ._lifecycle import ContextHandleMixin, WarnUnclosedMixin
from dataclasses import dataclass
from typing import Any, Callable

_REQUEST_HANDLE_CREATE_KEY = object()


class ResourceKind(UnknownIntEnum):
    """Network resource kind passed to resource callbacks."""

    UNKNOWN = 0
    STYLE = 1
    SOURCE = 2
    TILE = 3
    GLYPHS = 4
    SPRITE_IMAGE = 5
    SPRITE_JSON = 6
    IMAGE = 7


class ResourceLoadingMethod(UnknownIntEnum):
    """Resource loading method requested by MapLibre Native."""

    ALL = 0
    CACHE_ONLY = 1
    NETWORK_ONLY = 2


class ResourcePriority(UnknownIntEnum):
    """Resource request priority."""

    REGULAR = 0
    LOW = 1


class ResourceUsage(UnknownIntEnum):
    """Resource request usage."""

    ONLINE = 0
    OFFLINE = 1


class ResourceStoragePolicy(UnknownIntEnum):
    """Resource cache storage policy."""

    PERMANENT = 0
    VOLATILE = 1


class ResourceResponseStatus(NativeIntEnum):
    """Status for a resource provider response."""

    OK = 0
    ERROR = 1
    NO_CONTENT = 2
    NOT_MODIFIED = 3


class ResourceErrorReason(UnknownIntEnum):
    """Resource error reason used in provider responses and events."""

    NONE = 0
    NOT_FOUND = 1
    SERVER = 2
    CONNECTION = 3
    RATE_LIMIT = 4
    OTHER = 5


class ResourceProviderDecision(NativeIntEnum):
    """Decision returned by a resource provider callback."""

    PASS_THROUGH = 0
    HANDLE = 1


@dataclass(frozen=True, slots=True)
class ByteRange:
    """Byte range requested for a network resource."""

    start: int
    end: int

    @classmethod
    def _from_native(cls, raw: dict[str, Any]) -> "ByteRange":
        """Build a byte range from private native values."""
        return cls(start=raw["start"], end=raw["end"])


@dataclass(frozen=True, slots=True)
class ResourceRequest:
    """Copied request passed to a resource provider callback."""

    url: str
    kind: ResourceKind
    loading_method: ResourceLoadingMethod
    priority: ResourcePriority
    usage: ResourceUsage
    storage_policy: ResourceStoragePolicy
    range: ByteRange | None
    prior_modified_unix_ms: int | None
    prior_expires_unix_ms: int | None
    prior_etag: str | None
    prior_data: bytes

    @classmethod
    def _from_native(cls, raw: dict[str, Any]) -> "ResourceRequest":
        """Build a copied request from private native values."""
        raw_range = raw["range"]
        return cls(
            url=raw["url"],
            kind=ResourceKind(raw["kind"]),
            loading_method=ResourceLoadingMethod(raw["loading_method"]),
            priority=ResourcePriority(raw["priority"]),
            usage=ResourceUsage(raw["usage"]),
            storage_policy=ResourceStoragePolicy(raw["storage_policy"]),
            range=ByteRange._from_native(raw_range) if raw_range is not None else None,
            prior_modified_unix_ms=raw["prior_modified_unix_ms"],
            prior_expires_unix_ms=raw["prior_expires_unix_ms"],
            prior_etag=raw["prior_etag"],
            prior_data=raw["prior_data"],
        )


@dataclass(frozen=True, slots=True)
class ResourceTransformRequest:
    """Copied request passed to a resource transform callback."""

    kind: ResourceKind
    url: str

    @classmethod
    def _from_native(cls, raw: dict[str, Any]) -> "ResourceTransformRequest":
        """Build a copied transform request from private native values."""
        return cls(kind=ResourceKind(raw["kind"]), url=raw["url"])


@dataclass(frozen=True, slots=True)
class ResourceResponse:
    """Response used to complete a handled resource request."""

    status: ResourceResponseStatus = ResourceResponseStatus.OK
    error_reason: ResourceErrorReason = ResourceErrorReason.NONE
    bytes: bytes = b""
    error_message: str | None = None
    must_revalidate: bool = False
    modified_unix_ms: int | None = None
    expires_unix_ms: int | None = None
    etag: str | None = None
    retry_after_unix_ms: int | None = None

    def _to_native(self) -> dict[str, Any]:
        """Return private native bridge values for this response."""
        return {
            "status": self.status.native_code,
            "error_reason": self.error_reason.native_code,
            "bytes": self.bytes,
            "error_message": self.error_message,
            "must_revalidate": self.must_revalidate,
            "modified_unix_ms": self.modified_unix_ms,
            "expires_unix_ms": self.expires_unix_ms,
            "etag": self.etag,
            "retry_after_unix_ms": self.retry_after_unix_ms,
        }


class ResourceRequestHandle(WarnUnclosedMixin, ContextHandleMixin):
    """One-shot handle for a resource provider request selected for handling."""

    _handle_name = "ResourceRequestHandle"

    def __init__(self, native: Any, *, _create_key: object | None = None) -> None:
        if _create_key is not _REQUEST_HANDLE_CREATE_KEY:
            msg = "ResourceRequestHandle instances are created by resource providers"
            raise TypeError(msg)
        self._native = native
        self._closed = False

    @classmethod
    def _from_native(cls, native: Any) -> "ResourceRequestHandle":
        return cls(native, _create_key=_REQUEST_HANDLE_CREATE_KEY)

    @property
    def closed(self) -> bool:
        """Return whether this request handle has been completed or released."""
        return self._closed

    def complete(self, response: ResourceResponse) -> None:
        """Complete this request with a copied response."""
        if self._closed:
            from .errors import InvalidStateError

            raise InvalidStateError(
                None,
                "resource request handle is already closed",
            )
        native_response = response._to_native()
        self._native.validate_completion_response(native_response)
        try:
            self._native.complete(native_response)
        except BaseException:
            self.close()
            raise
        self._closed = True

    def is_cancelled(self) -> bool:
        """Return whether native code has cancelled the request."""
        if self._closed:
            from .errors import InvalidStateError

            raise InvalidStateError(
                None,
                "resource request handle is already closed",
            )
        return bool(self._native.is_cancelled())

    def close(self) -> None:
        """Release this request handle without completing it."""
        if self._closed:
            return
        self._native.close()
        self._closed = True


ResourceTransformCallback = Callable[[ResourceTransformRequest], str | None]
ResourceProviderCallback = Callable[
    [ResourceRequest, ResourceRequestHandle], ResourceProviderDecision
]


def _adapt_resource_transform_callback(
    callback: ResourceTransformCallback,
) -> Callable[[dict[str, Any]], str | None]:
    """Adapt a public resource transform callback for the native bridge."""

    def adapted(raw_request: dict[str, Any]) -> str | None:
        return callback(ResourceTransformRequest._from_native(raw_request))

    return adapted


def _adapt_resource_provider_callback(
    callback: ResourceProviderCallback,
) -> Callable[[dict[str, Any], Any], int]:
    """Adapt a public resource provider callback for the native bridge."""

    def adapted(raw_request: dict[str, Any], native_handle: Any) -> int:
        handle = ResourceRequestHandle._from_native(native_handle)  # noqa: SLF001
        try:
            decision = ResourceProviderDecision(
                callback(ResourceRequest._from_native(raw_request), handle)
            )
        except BaseException:
            if not handle.closed:
                handle.close()
            raise
        if decision is not ResourceProviderDecision.HANDLE and not handle.closed:
            handle.close()
        return decision.native_code

    return adapted


__all__ = [
    "ByteRange",
    "ResourceErrorReason",
    "ResourceKind",
    "ResourceLoadingMethod",
    "ResourcePriority",
    "ResourceProviderCallback",
    "ResourceProviderDecision",
    "ResourceRequest",
    "ResourceRequestHandle",
    "ResourceResponse",
    "ResourceResponseStatus",
    "ResourceStoragePolicy",
    "ResourceTransformCallback",
    "ResourceTransformRequest",
    "ResourceUsage",
]

File: bindings/dart/lib/src/resource/resource.dart
/// Resource requests, responses, transforms, providers, and request handles.
library;

import 'dart:typed_data';

/// Resource kind requested by MapLibre Native.
final class ResourceKind {
  const ResourceKind._(this.rawValue, this.name);

  /// Unknown resource kind.
  static const unknown = ResourceKind._(0, 'unknown');

  /// Style document.
  static const style = ResourceKind._(1, 'style');

  /// Source metadata.
  static const source = ResourceKind._(2, 'source');

  /// Tile payload.
  static const tile = ResourceKind._(3, 'tile');

  /// Glyph payload.
  static const glyphs = ResourceKind._(4, 'glyphs');

  /// Sprite image.
  static const spriteImage = ResourceKind._(5, 'spriteImage');

  /// Sprite JSON.
  static const spriteJson = ResourceKind._(6, 'spriteJson');

  /// Image payload.
  static const image = ResourceKind._(7, 'image');

  /// Creates a resource kind from a raw native value.
  factory ResourceKind.fromRawValue(int rawValue) => switch (rawValue) {
    0 => unknown,
    1 => style,
    2 => source,
    3 => tile,
    4 => glyphs,
    5 => spriteImage,
    6 => spriteJson,
    7 => image,
    _ => ResourceKind._(rawValue, 'unknown($rawValue)'),
  };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Resource loading method.
final class ResourceLoadingMethod {
  const ResourceLoadingMethod._(this.rawValue, this.name);

  /// All loading methods.
  static const all = ResourceLoadingMethod._(0, 'all');

  /// Cache-only loading.
  static const cacheOnly = ResourceLoadingMethod._(1, 'cacheOnly');

  /// Network-only loading.
  static const networkOnly = ResourceLoadingMethod._(2, 'networkOnly');

  /// Creates a resource loading method from a raw native value.
  factory ResourceLoadingMethod.fromRawValue(int rawValue) =>
      switch (rawValue) {
        0 => all,
        1 => cacheOnly,
        2 => networkOnly,
        _ => ResourceLoadingMethod._(rawValue, 'unknown($rawValue)'),
      };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Resource priority.
final class ResourcePriority {
  const ResourcePriority._(this.rawValue, this.name);

  /// Regular priority.
  static const regular = ResourcePriority._(0, 'regular');

  /// Low priority.
  static const low = ResourcePriority._(1, 'low');

  /// Creates a resource priority from a raw native value.
  factory ResourcePriority.fromRawValue(int rawValue) => switch (rawValue) {
    0 => regular,
    1 => low,
    _ => ResourcePriority._(rawValue, 'unknown($rawValue)'),
  };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Resource usage.
final class ResourceUsage {
  const ResourceUsage._(this.rawValue, this.name);

  /// Online usage.
  static const online = ResourceUsage._(0, 'online');

  /// Offline usage.
  static const offline = ResourceUsage._(1, 'offline');

  /// Creates a resource usage from a raw native value.
  factory ResourceUsage.fromRawValue(int rawValue) => switch (rawValue) {
    0 => online,
    1 => offline,
    _ => ResourceUsage._(rawValue, 'unknown($rawValue)'),
  };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Resource storage policy.
final class ResourceStoragePolicy {
  const ResourceStoragePolicy._(this.rawValue, this.name);

  /// Permanent storage.
  static const permanent = ResourceStoragePolicy._(0, 'permanent');

  /// Volatile storage.
  static const volatile = ResourceStoragePolicy._(1, 'volatile');

  /// Creates a resource storage policy from a raw native value.
  factory ResourceStoragePolicy.fromRawValue(int rawValue) =>
      switch (rawValue) {
        0 => permanent,
        1 => volatile,
        _ => ResourceStoragePolicy._(rawValue, 'unknown($rawValue)'),
      };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Resource response status.
final class ResourceResponseStatus {
  const ResourceResponseStatus._(this.rawValue, this.name);

  /// Successful response.
  static const ok = ResourceResponseStatus._(0, 'ok');

  /// Error response.
  static const error = ResourceResponseStatus._(1, 'error');

  /// No-content response.
  static const noContent = ResourceResponseStatus._(2, 'noContent');

  /// Not-modified response.
  static const notModified = ResourceResponseStatus._(3, 'notModified');

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Resource error reason.
final class ResourceErrorReason {
  const ResourceErrorReason._(this.rawValue, this.name);

  /// No error.
  static const none = ResourceErrorReason._(0, 'none');

  /// Resource was not found.
  static const notFound = ResourceErrorReason._(1, 'notFound');

  /// Server returned an error.
  static const server = ResourceErrorReason._(2, 'server');

  /// Connection failed.
  static const connection = ResourceErrorReason._(3, 'connection');

  /// Request was rate-limited.
  static const rateLimit = ResourceErrorReason._(4, 'rateLimit');

  /// Other resource error.
  static const other = ResourceErrorReason._(5, 'other');

  /// Creates an error reason from a raw native value.
  factory ResourceErrorReason.fromRawValue(int rawValue) => switch (rawValue) {
    0 => none,
    1 => notFound,
    2 => server,
    3 => connection,
    4 => rateLimit,
    5 => other,
    _ => ResourceErrorReason._(rawValue, 'unknown($rawValue)'),
  };

  /// Raw native value.
  final int rawValue;

  /// Human-readable name.
  final String name;
}

/// Exact URL rewrite rule used by the runtime resource transform.
final class ResourceUrlRewriteRule {
  /// Creates an exact URL rewrite rule.
  const ResourceUrlRewriteRule({
    this.kind,
    required this.url,
    required this.replacementUrl,
  });

  /// Optional resource kind filter. Null matches any kind.
  final ResourceKind? kind;

  /// Original URL to match exactly.
  final String url;

  /// Replacement URL returned to native code.
  final String replacementUrl;
}

/// Copied resource request delivered to a Dart resource provider.
final class ResourceRequest {
  /// Creates a copied resource request.
  ResourceRequest({
    required this.requestedUrl,
    required this.resolvedUrl,
    required this.kind,
    required this.loadingMethod,
    required this.priority,
    required this.usage,
    required this.storagePolicy,
    this.range,
    this.priorModifiedUnixMs,
    this.priorExpiresUnixMs,
    this.priorEtag,
    Uint8List? priorData,
  }) : priorData = priorData == null
           ? null
           : Uint8List.fromList(priorData).asUnmodifiableView();

  /// URL entering the network layer, preserving configured scheme aliases.
  final String requestedUrl;

  /// URL to fetch, after tile server normalization.
  final String resolvedUrl;

  /// Requested resource kind.
  final ResourceKind kind;

  /// Loading method requested by native code.
  final ResourceLoadingMethod loadingMethod;

  /// Request priority.
  final ResourcePriority priority;

  /// Request usage.
  final ResourceUsage usage;

  /// Request storage policy.
  final ResourceStoragePolicy storagePolicy;

  /// Optional byte range as inclusive start/end values.
  final ({BigInt start, BigInt end})? range;

  /// Optional previously modified timestamp in Unix milliseconds.
  final int? priorModifiedUnixMs;

  /// Optional previously expires timestamp in Unix milliseconds.
  final int? priorExpiresUnixMs;

  /// Optional previous entity tag.
  final String? priorEtag;

  /// Optional previous data bytes copied from native code.
  final Uint8List? priorData;
}

/// Exact URL route used by a queued Dart resource provider callback.
final class ResourceProviderRoute {
  /// Creates an exact URL provider route.
  const ResourceProviderRoute({this.kind, required this.requestedUrl});

  /// Optional resource kind filter. Null matches any kind.
  final ResourceKind? kind;

  /// Requested URL to match exactly.
  final String requestedUrl;
}

/// Exact URL provider rule used by the runtime resource provider.
final class ResourceProviderRule {
  /// Creates an exact URL provider rule.
  const ResourceProviderRule({
    this.kind,
    required this.requestedUrl,
    required this.response,
  });

  /// Optional resource kind filter. Null matches any kind.
  final ResourceKind? kind;

  /// Requested URL to match exactly.
  final String requestedUrl;

  /// Response to complete for matching requests.
  final ResourceResponse response;
}

/// Resource response returned by a Dart resource provider.
final class ResourceResponse {
  /// Creates a resource response.
  ResourceResponse({
    required this.status,
    this.errorReason = ResourceErrorReason.none,
    Uint8List? bytes,
    this.errorMessage,
    this.mustRevalidate = false,
    this.modifiedUnixMs,
    this.expiresUnixMs,
    this.etag,
    this.retryAfterUnixMs,
  }) : bytes = bytes == null
           ? null
           : Uint8List.fromList(bytes).asUnmodifiableView();

  /// Response status.
  final ResourceResponseStatus status;

  /// Error reason for error responses.
  final ResourceErrorReason errorReason;

  /// Optional response bytes.
  final Uint8List? bytes;

  /// Optional error message.
  final String? errorMessage;

  /// Whether cached data must be revalidated.
  final bool mustRevalidate;

  /// Optional modified timestamp in Unix milliseconds.
  final int? modifiedUnixMs;

  /// Optional expiration timestamp in Unix milliseconds.
  final int? expiresUnixMs;

  /// Optional entity tag.
  final String? etag;

  /// Optional retry-after timestamp in Unix milliseconds.
  final int? retryAfterUnixMs;
}
